src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs

```diff
@@ -76,7 +76,18 @@ public FieldBuilder DefineField(string fieldName, Type type, FieldAttributes att
 
         public FieldBuilder DefineField(string fieldName, Type type, Type[]? requiredCustomModifiers, Type[]? optionalCustomModifiers,
             FieldAttributes attributes)
-                => DefineFieldCore(fieldName, type, requiredCustomModifiers, optionalCustomModifiers, attributes);
+        {
+            ArgumentException.ThrowIfNullOrEmpty(fieldName);
+            ArgumentNullException.ThrowIfNull(type);
+
+            if (fieldName[0] == '\0')
+                throw new ArgumentException(SR.Argument_IllegalName, nameof(fieldName));
+
+            if (type == typeof(void))
```

Comment:
The `void` type in the actual builder implementation may be represented by completely different Type instance. This kind of validation can be done in the actual builder implementation only. I would suggest keeping it in the runtime builder.

It is debatable whether it is a good idea to do this validation in Reflection.Emit implementation at all. There are other potentially invalid field types, and it does not make sense to try to check for all of them in Reflection.Emit. We have separate ILVerify tool for that.